### PR TITLE
feat(weight-free): weight-free ONNX export for causalLM

### PR DESCRIPTION
**Pull request by quic-amitraj**
## Summary

Adds weight-free ONNX export for causal LM models on top of the dynamo export infrastructure.

Weight-free export builds the model on meta device (no weights in RAM), exports ONNX graph structure only, and lets the QAIC compiler load weights directly from the original safetensors checkpoint at compile time via `weight_spec.json`.

  ## New files

- `QEfficient/exporter/weight_free/` — core weight-free export, weight spec format, checkpoint prep pipeline (MoE expert stacking, dtype
  conversion, .bin→safetensors auto-convert)
- `examples/text_generation/weight_free/` — single-model smoke test, CB weight-free smoke test

  ## Key changes

- `modeling_qeff.py` — `use_weight_free_export` flag; reorder `example_inputs`/`dynamic_shapes` by `forward()` signature so `torch.export`
binds them correctly; skip `SplitTensorsTransform`; embed `weight_spec.json` as ONNX metadata; symlink prepared checkpoint next to ONNX
- `modeling_auto.py` — `use_weight_free_export` wired through `export()` and `compile()`; auto-enable `dynamo=True`
- `onnx_transforms.py` — disable unsupported transforms for dynamo path
- `export_utils.py` — separate dynamo/TorchScript subfunction setup paths
- `torch_patches.py` — `temporarily_enable/disable_nested_compile_regions()`
- `core.py` — dtype suffix in prepared checkpoint dir; prune fake initializers after meta-device export

  ## Model fixes

- `codegen` — fix `embed_positions` device/dtype for weight-free subfunction tracing
- `glm4_moe` — `logits.float()` to fix runtime buffer mismatch; rotary cos/sin device sync; `_rotary_dim` attribute; MoE expert placeholder
  dtype
- `granitemoe`, `mixtral`, `phi3`, `qwen3_moe` — rotary cos/sin device/dtype sync for meta-device tracing; MoE expert placeholder dtype
  - `continuous_batching.py` — fix `use_dynamo=True` → `dynamo=True`

  ## Tests

  ```bash
  pip install -e .[test]
  pytest tests/weight_free/ -m "not on_qaic" -n auto -v

  ```

---------

### CODE REVIEW

**Comment by quic-amitraj** [comment]
CI-Ready

**Comment by quic-amitraj** [comment]
CI-Ready

File: docs/source/features_enablement.md [hunk 1]
```diff
@@ -39,6 +39,45 @@ python -m QEfficient.cloud.infer --model_name gpt2 --batch_size 1 --prompt_len 3
...
+    prefill_seq_len=128,
+    ctx_len=128,
+    num_cores=16,
+    dynamo=True,
+    use_onnx_subfunctions=True,
+    use_weight_free_export=True,
+)
+```
+
+`use_weight_free_export=True` forces the dynamo export path. When constructing a
```
Comment (ochougul) [inline docs/source/features_enablement.md]: didn't we change this to `weight_free` ?

File: examples/dynamo/causal_lm/basic_dynamo_inference.py [hunk 2]
```diff
@@ -42,14 +52,17 @@ def main():
...
     if args.num_hidden_layers > 0:
         config.num_hidden_layers = args.num_hidden_layers
-
-    # Load model and apply QEff transforms
-    model = QEFFAutoModelForCausalLM.from_pretrained(args.model_name, config=config)
+    model = QEFFAutoModelForCausalLM.from_pretrained(
+        args.model_name,
+        config=config,
+        weight_free=args.weight_free,
+        enable_proxy=args.enable_proxy,
```
Comment (ochougul) [inline examples/dynamo/causal_lm/basic_dynamo_inference.py]: why `enable_proxy` here?

File: QEfficient/base/checkpoint_transforms.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        **kwargs,
+    ) -> Path:
+        """Apply the first matching transform and return the usable checkpoint directory."""
+        src, out = Path(src), Path(out)
+
+        source_dir = src
+        has_safetensors = bool(list(src.glob("*.safetensors"))) or (src / "model.safetensors.index.json").exists()
+        if not has_safetensors and list(src.glob("*.bin")):
+            source_dir = out.with_name(out.name + "-source-safetensors")
+            convert_bin_to_safetensors(src, source_dir)
```
Comment (ochougul) [inline QEfficient/base/checkpoint_transforms.py:168]: for bin files we do an extra write of all the weights?

File: QEfficient/transformers/models/modeling_auto.py [hunk 4]
```diff
@@ -188,7 +195,17 @@ def _build_meta_model(hf_auto_class, pretrained_model_name_or_path, kwargs):
...
+    Uses accelerate's ``init_empty_weights()`` rather than a raw
+    ``torch.device("meta")`` context: some models (e.g. Gemma's embed_scale)
+    compute non-persistent buffers from a plain Python constant inside
+    ``__init__`` via ``register_buffer(..., persistent=False)``. A bare
+    ``torch.device("meta")`` context turns even those constant-valued buffers
+    into empty meta tensors with no data, which later fails to serialize to
+    ONNX. ``init_empty_weights()`` only intercepts parameter/persistent-buffer
+    allocation, so such computed buffers still get their real values.
     """
+    from accelerate import init_empty_weights
```
Comment (ochougul) [inline QEfficient/transformers/models/modeling_auto.py:209]: add warning here to install req file

File: QEfficient/transformers/models/modeling_auto.py [hunk 5]
```diff
@@ -4149,6 +4215,7 @@ def _legacyify_cache(obj):
                 use_onnx_subfunctions=kwargs.get("use_onnx_subfunctions", False),
                 dynamo=dynamo,
                 offload_pt_weights=kwargs.get("offload_pt_weights", True),
+                prefill_only=prefill_only,
```
Comment (ochougul) [inline QEfficient/transformers/models/modeling_auto.py]: remove?

File: QEfficient/base/modeling_qeff.py [hunk 6]
```diff
@@ -297,7 +299,7 @@ def model_name(self) -> str:
 
     @property
     @abstractmethod
-    def get_model_config(self) -> Dict:
+    def get_model_config(self) -> dict:
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: revert

File: QEfficient/base/modeling_qeff.py
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: remove this and `prefill_only` as well, its not gettting used

File: QEfficient/base/modeling_qeff.py [hunk 7]
```diff
@@ -573,17 +496,39 @@ def _resolve_pkv_names(layer_idx, layer_state):
...
                     export_kwargs,
                 )
+                self._offload_model_weights(offload_pt_weights)
             else:
-                self._export_via_legacy(
+                from QEfficient.exporter.onnx_exporter import export_via_legacy
+
+                self._model_offloaded_check()
+                export_result = export_via_legacy(
+                    self,
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py:526]: `self._offload_model_weights(offload_pt_weights)` should be called after this in the else condition

File: QEfficient/exporter/onnx_exporter.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    weight_spec_path = final_weight_spec if final_weight_spec.exists() else None
+
+    post_transform_hooks = ()
+    post_export_hooks = ()
+    if weight_spec_path is not None:
+
+        def _embed_weight_spec(model, spec_path=weight_spec_path):
+            from QEfficient.exporter.weight_free.export import embed_weight_spec_as_metadata
+
+            embed_weight_spec_as_metadata(model, spec_path)
```
Comment (ochougul) [inline QEfficient/exporter/onnx_exporter.py]: better to use `partial`

File: QEfficient/base/modeling_qeff.py [hunk 9]
```diff
@@ -592,33 +537,40 @@ def _resolve_pkv_names(layer_idx, layer_state):
...
                 _restore_retained_state_output_names(model, output_names)
 
-            # Add metadata to the model
+            transform_names = [transform.__name__ for transform in self._pytorch_transforms + active_transforms]
             model.metadata_props.append(
-                onnx.StringStringEntryProto(key="qeff_transforms", value=",".join(self._transform_names()))
+                onnx.StringStringEntryProto(key="qeff_transforms", value=",".join(transform_names))
             )
+            for hook in export_result.post_transform_hooks:
+                hook(model)
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: not needed it can be a simple method inside the if weight_free condition

File: QEfficient/base/modeling_qeff.py [hunk 10]
```diff
@@ -628,6 +580,9 @@ def _resolve_pkv_names(layer_idx, layer_state):
             gc.collect()
             logger.info("Transformed ONNX saved")
 
+            for hook in export_result.post_export_hooks:
+                hook()
+
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: can be simple method call in weightfree

File: QEfficient/transformers/models/modeling_auto.py (hunk 4 above)
Comment (quic-amitraj) [reply QEfficient/transformers/models/modeling_auto.py:209]: Added in from_preatrained.

File: QEfficient/exporter/weight_free/checkpoint_transforms.py [hunk 11]
```diff
@@ -0,0 +1,998 @@
...
+        **kwargs,
+    ) -> bool:
+        """Convert checkpoint shards to ``target_dtype`` in a prepared output directory."""
+        sentinel = out / _SENTINEL
+        if sentinel.exists():
+            logger.info("DtypeConversionCheckpointTransform: prepared checkpoint exists, skipping.")
+            return False
+
+        out.mkdir(parents=True, exist_ok=True)
+        copy_checkpoint_aux_files(src, out)
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/checkpoint_transforms.py:224]: This is independent of data type conversion why is it inside this transform. keep it outside as a separate transform

File: QEfficient/utils/checkpoint_utils.py [hunk 12]
```diff
@@ -23,3 +35,258 @@ def load_checkpoint(model, checkpoint: str, strict=False, post_process_func=None
...
+        return json.loads(index_path.read_text())["weight_map"]
+    shard_files = sorted(src.glob("*.safetensors"))
+    if not shard_files:
+        raise FileNotFoundError(f"No safetensors files found in {src}")
+    weight_map: Dict[str, str] = {}
+    for sf in shard_files:
+        with safe_open(str(sf), framework="pt") as f:
+            for k in f.keys():
+                weight_map[k] = sf.name
+    return weight_map
```
Comment (ochougul) [inline QEfficient/utils/checkpoint_utils.py:103]: why are we reading the full safetensors if the `model.safetensors.index.json` is not present?
It should always present if the `saftetensors` file is present, otherwise we error out.

File: QEfficient/exporter/weight_free/checkpoint_transforms.py [hunk 13]
```diff
@@ -0,0 +1,998 @@
...
+            weight_map: Dict[str, str] = json.loads(index_path.read_text())["weight_map"]
+        else:
+            shards = sorted(src.glob("*.safetensors"))
+            if not shards:
+                return False
+            weight_map = {}
+            for shard in shards:
+                with safe_open(str(shard), framework="pt") as f:
+                    for k in f.keys():
+                        weight_map[k] = shard.name
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/checkpoint_transforms.py:804]: wrong, and duplicate of `read_weight_map`

File: QEfficient/base/checkpoint_transforms.py [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+        Returns True if the checkpoint was prepared, False if skipped (idempotent)."""
+        raise NotImplementedError
+
+    @classmethod
+    def is_applicable(cls, weight_map: Dict[str, str], **kwargs) -> bool:
+        """Return True if this transform should run for the given checkpoint."""
+        return True
+
+
+class CheckpointTransformPipeline:
```
Comment (ochougul) [inline QEfficient/base/checkpoint_transforms.py:126]: rename to `CheckpointTransformSelector` it does not pipeline anything.
Also, we read the metadata here and again read it in the actual transform, that can be avoided.

Comment (ochougul) [reply QEfficient/base/checkpoint_transforms.py:126]: I think this design should be reworked before we build more on it. Right now CheckpointTransformPipeline is really a first-match selector: it picks one transform, and that transform owns reading, processing, and writing the whole checkpoint.

  A better shape would be: decide all applicable transforms up front from model/config/quantization metadata, then stream each weight through the selected transform steps and write the final checkpoint once. Regex over safetensor keys is fine inside a selected transform to route tensors, but it should not be the primary signal for deciding which model-level transform applies, since we do not control checkpoint key naming.

Comment (ochougul) [reply QEfficient/base/checkpoint_transforms.py:126]: decode applicability of transform based on regex key matching is wrong, we can depend on model architecture.

File: QEfficient/exporter/weight_free/checkpoint_key_resolver.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+
+    for name, init_value in list(model_ir.graph.initializers.items()):
+        if name not in model_names:
+            continue
+
+        onnx_name = tied_weight_map.get(name, name)
+        checkpoint_key = find_checkpoint_key(onnx_name, checkpoint_index, backbone)
+        if checkpoint_key is None:
+            # Computed buffers such as rotary caches are left embedded in ONNX.
+            continue
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/checkpoint_key_resolver.py]: For model params/buffers, missing safetensor resolution should be an error unless the name is explicitly classified as a computed buffer such as RoPE cache. Otherwise weight-free export can silently embed or drop a real weight.

File: QEfficient/exporter/weight_free/checkpoint_key_resolver.py [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+        for alias in _moe_weight_aliases(candidate):
+            if alias in seen:
+                continue
+            seen.add(alias)
+            if alias in checkpoint_index:
+                return alias
+    return None
+
+
+def find_checkpoint_key(
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/checkpoint_key_resolver.py:107]: find_checkpoint_key() is deterministic but too heuristic: it tries a chain of prefix and MoE alias fallbacks and accepts the first match. If multiple candidate keys exist, we silently choose based on hardcoded order rather than model/config layout. This should either detect ambiguity and fail, or receive an explicit model/layout mapping from the checkpoint transform/selector.

File: QEfficient/exporter/onnx_exporter.py [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+        export_kwargs=wf_export_kwargs,
+        onnx_transform_kwargs=onnx_transform_kwargs or {},
+    )
+
+    final_weight_spec = onnx_path.with_name("weight_spec.json")
+    weight_spec_path = final_weight_spec if final_weight_spec.exists() else None
+
+    post_transform_hooks = ()
+    post_export_hooks = ()
+    if weight_spec_path is not None:
```
Comment (ochougul) [inline QEfficient/exporter/onnx_exporter.py]: remove this if condition, we should always expect a valid json file here

File: QEfficient/exporter/weight_free/export.py [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+    Used to embed weight_spec.json into the ONNX so the QAIC compiler
+    can locate external weight files without a separate sidecar lookup.
+    """
+    import onnx
+
+    for entry in model.metadata_props:
+        if entry.key == key:
+            entry.value = value
+            return
+    model.metadata_props.append(onnx.StringStringEntryProto(key=key, value=value))
```
Comment (ochougul) [inline QEfficient/exporter/weight_free/export.py]: not needed. slop

**Review by ochougul — CHANGES_REQUESTED** [review]

File: .github/workflows/unit-tests.yml [hunk 19]
```diff
@@ -58,3 +58,38 @@
           pytest tests/dynamo/ \
             -m "not on_qaic" \
             -n auto -v
+
+  weight-free-tests:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
```
[the remark this answers is not in the document]
Comment (ochougul) [reply .github/workflows/unit-tests.yml:66]: Can this warning be fixed?

**Comment by quic-amitraj** [comment]
CI-Ready
